### PR TITLE
Cherry-pick hierarchy paths SQL change

### PR DESCRIPTION
**Pull request by oliverguenther**
https://github.com/opf/openproject/pull/6825

/cc @steffzt

### CODE REVIEW

**Comment by oliverguenther** [comment]
Failing spec is unrelated.

File: app/models/relation/hierarchy_paths.rb [hunk 1]
```diff
@@ -86,21 +86,49 @@ def alter_hierarchy_path
...
-          (work_package_id, path)
-        SELECT
-          to_id, #{add_hierarchy_agg_function} AS path
-          FROM
-          (SELECT to_id, from_id, hierarchy
-          FROM relations
-          WHERE hierarchy > 0 AND relates = 0 AND blocks = 0 AND duplicates = 0 AND includes = 0 AND requires = 0 AND follows = 0
-          #{add_hierarchy_id_constraint(id)}
-          ) ordered_by_hierarchy
+      # Added a slighly optimized sql statement, which works without subselects. STEFFEN
```
Comment (ulferts) [inline app/models/relation/hierarchy_paths.rb:89]: Do not describe what has been done here but rather the purpose of those following lines of code. We also do not use Dev's names in comments

File: app/models/relation/hierarchy_paths.rb [hunk 2]
```diff
@@ -86,21 +86,49 @@ def alter_hierarchy_path
...
+            #{hierarchy_table_name}
+            (work_package_id, path)
+          SELECT
+            to_id, #{add_hierarchy_agg_function} AS path
+          FROM (
+            SELECT to_id, from_id, hierarchy
+            FROM relations
+            WHERE to_id = #{id} AND
+                  hierarchy > 0 AND relates = 0 AND blocks = 0 AND duplicates = 0 AND includes = 0 AND requires = 0 AND follows = 0
+            UNION SELECT to_id,from_id,hierarchy
```
Comment (ulferts) [inline app/models/relation/hierarchy_paths.rb:117]: whitespace

File: app/models/relation/hierarchy_paths.rb [hunk 3]
```diff
@@ -86,21 +86,49 @@ def alter_hierarchy_path
...
+                  FROM relations
+                  WHERE from_id=#{id} AND
+                        hierarchy > 0 AND relates = 0 AND blocks = 0 AND duplicates = 0 AND includes = 0 AND requires = 0 AND follows = 0
+            UNION SELECT b.to_id, b.from_id, b.hierarchy FROM relations a
+                  JOIN relations b ON b.to_id = a.to_id
+                  WHERE a.from_id = #{id} AND
+                        a.hierarchy > 0 AND a.relates = 0 AND a.blocks = 0 AND a.duplicates = 0 AND a.includes = 0 AND a.requires = 0 AND a.follows = 0 AND
+                        b.hierarchy > 0 AND b.relates = 0 AND b.blocks = 0 AND b.duplicates = 0 AND b.includes = 0 AND b.requires = 0 AND b.follows = 0
+          ) AS Foo
```
Comment (ulferts) [inline app/models/relation/hierarchy_paths.rb:126]: a more descriptive alias would probably be easy to find ;)

File: app/models/relation/hierarchy_paths.rb [hunk 4]
```diff
@@ -86,21 +86,49 @@ def alter_hierarchy_path
...
-        INSERT INTO
-          #{hierarchy_table_name}
-          (work_package_id, path)
-        SELECT
-          to_id, #{add_hierarchy_agg_function} AS path
-          FROM
-          (SELECT to_id, from_id, hierarchy
-          FROM relations
-          WHERE hierarchy > 0 AND relates = 0 AND blocks = 0 AND duplicates = 0 AND includes = 0 AND requires = 0 AND follows = 0
-          #{add_hierarchy_id_constraint(id)}
```
Comment (ulferts) [inline app/models/relation/hierarchy_paths.rb:99]: The method called `add_hierarchy_id_constraint` can be removed now.

File: app/models/relation/hierarchy_paths.rb [hunk 5]
```diff
@@ -86,21 +86,49 @@ def alter_hierarchy_path
...
+        SQL
+      else
+        stmt = <<-SQL
+          INSERT INTO
+            #{hierarchy_table_name}
+            (work_package_id, path)
+          SELECT
+            to_id, #{add_hierarchy_agg_function} AS path
+          FROM (
+            SELECT to_id, from_id, hierarchy
```
Comment (ulferts) [inline app/models/relation/hierarchy_paths.rb:113]: As the statements between having an id and not having only differs within the subselect, it makes sense to extract only those differences into a separate method.

**Review by ulferts — CHANGES_REQUESTED** [review]
The performance gain on MySql is considerable, so good work @steffzt. 

I have a few requests for changes in the code, though.

**Comment by ulferts** [comment]
@oliverguenther I can make the necessary changes to get the code to become mergeable but it would be nice, if @steffzt would be attributed with the code changes. I might change that as well.

**Comment by oliverguenther** [comment]
Closing as discussed, please force push your own PR or open a new one so the commit is correctly attributed to you! Thanks 🍻

File: app/models/relation/hierarchy_paths.rb (hunk 3 above)
Comment (steffzt) [reply app/models/relation/hierarchy_paths.rb:126]: Bar? ;-)

File: app/models/relation/hierarchy_paths.rb (hunk 1 above)
Comment (steffzt) [reply app/models/relation/hierarchy_paths.rb:89]: Is removed

File: app/models/relation/hierarchy_paths.rb (hunk 4 above)
Comment (steffzt) [reply app/models/relation/hierarchy_paths.rb:99]: done

File: app/models/relation/hierarchy_paths.rb (hunk 5 above)
Comment (steffzt) [reply app/models/relation/hierarchy_paths.rb:113]: yep, done-

**Pull request closed without merging**